renderdoc/driver/ihv/intel/intel_gl_counters.cpp

```diff
@@ -0,0 +1,309 @@
+  GLuint len;
+  uint32_t nSamples = m_glQueries.size() / (m_passIndex + 1);
+
+  for(uint32_t q = nSamples * m_passIndex; q < m_glQueries.size(); q++)
+  {
+    GL.glGetPerfQueryDataINTEL(m_glQueries[q], GL_PERFQUERY_WAIT_INTEL, data.size(), &data[0], &len);
+  }
+}
+
+void IntelGlCounters::BeginSample(uint32_t sampleID)
+{
+  GLuint queryId = m_EnabledQueries[m_passIndex];
+  GLuint queryHandle = 0;
+
+  GL.glCreatePerfQueryINTEL(queryId, &queryHandle);
+  m_glQueries.push_back(queryHandle);
+
+  GLenum err = GL.glGetError();
+  if(err != eGL_NONE)
+    return;
+
+  GL.glBeginPerfQueryINTEL(m_glQueries.back());
+}
+
+void IntelGlCounters::EndSample()
+{
+  GLuint queryHandle = m_glQueries.back();
+
+  if(queryHandle == 0)
+    return;
+
+  GL.glEndPerfQueryINTEL(queryHandle);
+}
+
+uint32_t IntelGlCounters::CounterPass(const IntelGlCounter &counter)
+{
+  for(uint32_t p = 0; p < m_EnabledQueries.size(); p++)
+    if(m_EnabledQueries[p] == counter.queryId)
+      return p;
+
+  RDCASSERT(false && "Counter not enabled");
+  return 0;
+}
+
+void IntelGlCounters::CopyData(void *dest, const IntelGlCounter &counter, uint32_t sample,
+                               uint32_t maxSampleIndex)
+{
+  uint32_t pass = CounterPass(counter);
+  uint32_t queryHandle = m_glQueries[maxSampleIndex * pass + sample];
+
+  std::vector<uint8_t> data(m_Queries[m_EnabledQueries[pass]].size);
+  GLuint len;
+  GL.glGetPerfQueryDataINTEL(queryHandle, 0, data.size(), &data[0], &len);
+
+  memcpy(dest, &data[counter.offset], counter.desc.resultByteWidth);
+}
+
+std::vector<CounterResult> IntelGlCounters::GetCounterData(uint32_t maxSampleIndex,
+                                                           const std::vector<uint32_t> &eventIDs,
+                                                           const std::vector<GPUCounter> &counters)
+{
+  std::vector<CounterResult> ret;
+
+  RDCASSERT((maxSampleIndex * m_EnabledQueries.size()) == m_glQueries.size());
+
+  for(uint32_t s = 0; s < maxSampleIndex; s++)
+  {
+    for(const GPUCounter &c : counters)
+    {
+      const IntelGlCounter &counter = m_Counters[GPUCounterToCounterIndex(c)];
+      switch(counter.desc.resultType)
+      {
+        case CompType::Double:
+        {
+          double r;
+          CopyData(&r, counter, s, maxSampleIndex);
+          ret.push_back(CounterResult(eventIDs[s], counter.desc.counter, r));
+          break;
+        }
+        case CompType::Float:
+        {
+          float r;
+          CopyData(&r, counter, s, maxSampleIndex);
+          ret.push_back(CounterResult(eventIDs[s], counter.desc.counter, r));
+          break;
+        }
+        case CompType::UInt:
+        {
+          uint64_t r;
+          CopyData(&r, counter, s, maxSampleIndex);
+          ret.push_back(CounterResult(eventIDs[s], counter.desc.counter, r));
+          break;
+        }
+        default:
+          RDCASSERT(false && "Wrong type");    // unreachable
```

Comment:
Still a couple of cases of (ab)using asserts as error messages here, FYI - see `IntelGlCounters::CounterPass` as well.

Reply:
Oops, fixed.